File: src/types/file.ts
/**
 * Идентификатор файла в формате name@version
 */
export type FileId = string;

/**
 * Модель файла/мода в системе
 *
 * Файл идентифицируется комбинацией name@version (как в npm пакетах).
 * Множественные версии одного файла могут сосуществовать.
 */
export interface File {
  id: number;
  name: string;
  version: string;
  path?: string;
  metadata: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

/**
 * Создать идентификатор файла из имени и версии
 */
export function createFileId(name: string, version: string): FileId {
  return `${name}@${version}`;
}

/**
 * Парсить идентификатор файла на имя и версию
 */
export function parseFileId(fileId: FileId): { name: string; version: string } | null {
  const parts = fileId.split('@');
  if (parts.length !== 2) {
    return null;
  }
  return { name: parts[0], version: parts[1] };
}

File: src-tauri/src/models/file.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Идентификатор файла в формате name@version
#[allow(dead_code)]
pub type FileId = String;

/// Структура данных для файла/мода
///
/// Представляет файл в системе с идентификацией через name@version.
/// Множественные версии одного файла могут сосуществовать.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct File {
    pub id: i64,
    pub name: String,
    pub version: String,
    pub path: Option<String>,
    #[serde(default)]
    pub metadata: serde_json::Value,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl File {
    /// Получить идентификатор файла в формате name@version
    #[allow(dead_code)]
    pub fn file_id(&self) -> FileId {
        format!("{}@{}", self.name, self.version)
    }

    /// Создать FileId из имени и версии
    #[allow(dead_code)]
    pub fn create_file_id(name: &str, version: &str) -> FileId {
        format!("{}@{}", name, version)
    }

    /// Парсить FileId на имя и версию
    #[allow(dead_code)]
    pub fn parse_file_id(file_id: &str) -> Option<(String, String)> {
        file_id
            .split_once('@')
            .map(|(name, version)| (name.to_string(), version.to_string()))
    }
}
